lib/upipe-modules/upipe_separate_fields.c

```diff
@@ -0,0 +1,364 @@
+        case UPIPE_SET_FLOW_DEF: {
+            struct uref *flow_def = va_arg(args, struct uref *);
+            return upipe_separate_fields_set_flow_def(upipe, flow_def);
+        }
+        case UPIPE_GET_FLOW_DEF:
+        case UPIPE_GET_OUTPUT:
+        case UPIPE_SET_OUTPUT:
+            return upipe_separate_fields_control_output(upipe, command, args);
+        default:
+            return UBASE_ERR_UNHANDLED;
+    }
+}
+
+static void upipe_separate_fields_free(struct upipe *upipe)
+{
+    upipe_separate_fields_clean_ubuf_mgr(upipe);
+    upipe_separate_fields_clean_urefcount(upipe);
+    upipe_separate_fields_clean_output(upipe);
+    upipe_separate_fields_clean_input(upipe);
+    upipe_separate_fields_free_void(upipe);
+}
+
+static struct upipe *upipe_separate_fields_alloc(struct upipe_mgr *mgr,
+                                           struct uprobe *uprobe,
+                                           uint32_t signature,
+                                           va_list args)
+{
+    struct upipe *upipe =
+        upipe_separate_fields_alloc_void(mgr, uprobe, signature, args);
+    if (unlikely(upipe == NULL))
+        return NULL;
+
+    upipe_separate_fields_init_urefcount(upipe);
+    upipe_separate_fields_init_input(upipe);
+    upipe_separate_fields_init_ubuf_mgr(upipe);
+    upipe_separate_fields_init_output(upipe);
+
+    return upipe;
+}
+
+static void upipe_separate_fields_input(struct upipe *upipe, struct uref *uref,
+                                  struct upump **upump_p)
+{
+    if (!upipe_separate_fields_check_input(upipe)) {
+        upipe_separate_fields_hold_input(upipe, uref);
+        upipe_separate_fields_block_input(upipe, upump_p);
+    } else if (!upipe_separate_fields_handle(upipe, uref, upump_p)) {
+        upipe_separate_fields_hold_input(upipe, uref);
+        upipe_separate_fields_block_input(upipe, upump_p);
+        /* Increment upipe refcount to avoid disappearing before all packets
+         * have been sent. */
+        upipe_use(upipe);
+    }
+}
+
+static bool upipe_separate_fields_handle(struct upipe *upipe, struct uref *uref,
+                                  struct upump **upump_p)
+{
+    struct upipe_separate_fields *ctx = upipe_separate_fields_from_upipe(upipe);
+    if (!ctx->ubuf_mgr)
+        return false;
+
+    size_t width, height;
+    uint8_t macropixel;
+    if (unlikely(!ubase_check(uref_pic_size(uref, &width, &height,
+                        &macropixel)))) {
+        upipe_warn(upipe, "dropping picture");
+        upipe_throw_error(upipe, UBASE_ERR_INVALID);
+        uref_free(uref);
+        return true;
+    }
+
+    bool has_progressive_attr = ubase_check(uref_pic_get_progressive(uref));
+    bool has_tff_attr         = ubase_check(uref_pic_get_tff(uref));
+
+    if (has_progressive_attr)
+        upipe_warn(upipe, "picture marked as progrssive, separating fields anyway");
+
+    /* first field */
+    struct ubuf *ubuf = ubuf_pic_alloc(ctx->ubuf_mgr, width, height / 2);
```

Comment:
Technically it could work with zero copy by changing the origin or end, and changing the stride. Changing the stride would require patching the manager but it should be easy.

Reply:
OK need to do this at some point, I wonder if I should split the PR in 2 modules then